update DRA api for additional usecases (#1)
rthallisey/community-1: design-proposals/dra.md

DRA allows vendors fine-grained control of devices. The device-plugin model will continue
to exist in Kubernetes, but DRA will offer vendors more control over the device topology.

## Motivation

DRA adoption is important for KubeVirt so that vendors can expect the same
control of their devices using Virtual Machines and Containers.

## Goals

- Align on how KubeVirt will consume external and in-tree DRA drivers
- Align on what drivers KubeVirt will support in tree
- Align on the API changes needed to consume DRA enabled devices in KubeVirt

## Non Goals

- Replace device-plugin support in KubeVirt

## Definition of Users

- A user is a person that wants to attach a device to a VM
- An admin is a person who manages the infrastructure and decide what kind of devices will be exposed via DRA
- An developer is a person familiar with CNCF ecosystem that can develop automation using the APIs discussed here

## User Stories

- As a user, I would like to use my GPU dra driver with KubeVirt
- As a user, I would like to use KubeVirt's default driver
- As a developer, I would like APIs to be extensible so I can develop drivers/webhooks/automation for custom use-cases

> ## Use Cases

Author: Create an assumption section dependencies (DRA, k8s version etc)

### Supported Usecases

1. Devices where the DRA driver that set agreed upon env variable will be supported. For example,
   `PCI_RESOURCE_<RESOURCE_CLAIM_NAME>` for a pci device with value being the bus id.
   1. There will not be any other restrictions on the DRA driver authors, they are free to implement their own schema
      of ClaimParameters, NodeAllocationState etc.

> 1. More than one device could be provisioned for a single resource claim. If this is the case, all devices must be

Author: This needs to change for 1.31

1. API changes
2. Driver Implementation

## API Changes

```
type VirtualMachineInstanceSpec struct {
    // ResourceClaims defines which ResourceClaims must be allocated
    // and reserved before the virt-launcher pod and hence the VMI is allowed to start. The resources
    // will be made available to the domain which consume them
    // by name.
    //
    // This is an alpha field and requires enabling the
    // DynamicResourceAllocation feature gate in kubernetes
    //  https://kubernetes.io/docs/concepts/scheduling-eviction/dynamic-resource-allocation/
    //
    // This field is immutable.
    //
    // +featureGate=DynamicResourceAllocation
    ResourceClaims []k8sv1.PodResourceClaim `json:"resourceClaims,omitempty"`
}

type GPU struct {
	// Name of the GPU device as exposed by a device plugin
	Name              string       `json:"name"`
	DeviceSource        DeviceSource `json:",inline"`
	VirtualGPUOptions *VGPUOptions `json:"virtualGPUOptions,omitempty"`
	// If specified, the virtual network interface address and its tag will be provided to the guest via config drive
	// +optional
	Tag   string `json:"tag,omitempty"`

> 	Claim string `json:"claim,omitempty"`

Reviewer: Is this meant to be removed? DeviceSource already has the ResourceClaim name

}

type HostDevice struct {
	Name string `json:"name"`
	// If specified, the virtual network interface address and its tag will be provided to the guest via config drive
	// +optional
	Tag string `json:"tag,omitempty"`
	DeviceSource        DeviceSource `json:",inline"`
}

type DeviceSource struct {
	// DeviceName is the name of the device provisioned by device-plugins
	DeviceName string `json:"deviceName,omitempty"`
	// ClaimName is the name of the claim that is going to provision the DRA device
    // +featureGate=DynamicResourceAllocation

> 	ClaimName string `json:"claimName,omitempty"`

Reviewer: nit: ResourceClaimName
```suggestion
	ResourceClaimName string `json:"resourceClaimName,omitempty"`
```

}

type VirtualMachineInstanceStatus struct {
	// Status of resource claims.
	// +featureGate=DynamicResourceAllocation
	// +optional

> 	ResourceClaimStatuses []VitualMachineInstanceResourceClaimStatus `json:"resourceClaimStatuses,omitempty"`

Reviewer: nit: ResourceClaimStatusList?

## Motivation

DRA adoption is important for KubeVirt so that vendors can expect the same
control of their devices using Virtual Machines and Containers.

## Goals

- Align on how KubeVirt will consume external and in-tree DRA drivers
- Align on what drivers KubeVirt will support in tree
- Align on the API changes needed to consume DRA enabled devices in KubeVirt

## Non Goals

- Replace device-plugin support in KubeVirt

## Definition of Users

- A user is a person that wants to attach a device to a VM
- An admin is a person who manages the infrastructure and decide what kind of devices will be exposed via DRA
- An developer is a person familiar with CNCF ecosystem that can develop automation using the APIs discussed here

## User Stories

- As a user, I would like to use my GPU dra driver with KubeVirt
- As a user, I would like to use KubeVirt's default driver
- As a developer, I would like APIs to be extensible so I can develop drivers/webhooks/automation for custom use-cases

## Use Cases

### Supported Usecases

1. Devices where the DRA driver that set agreed upon env variable will be supported. For example,

>    `PCI_RESOURCE_<RESOURCE_CLAIM_NAME>` for a pci device with value being the bus id.

Reviewer: ```suggestion
   `PCI_RESOURCE_<RESOURCE_CLAIM_NAME>` for a pci device with value being the PCIe bus address.
```

## Goals
- Align on how KubeVirt will consume external and in-tree DRA drivers
- Align on what drivers KubeVirt will support in tree
- Align on the API changes needed to consume DRA enabled devices in KubeVirt

## Non Goals
- Replace device-plugin support in KubeVirt

## Definition of Users
- A user is a person that wants to attach a device to a VM
- An admin is a person who manages the infrastructure and decide what kind of devices will be exposed via DRA
- An developer is a person familiar with CNCF ecosystem that can develop automation using the APIs discussed here  

## User Stories
- As a user, I would like to use my GPU dra driver with KubeVirt
- As a user, I would like to use KubeVirt's default driver
- As a developer, I would like APIs to be extensible so I can develop drivers/webhooks/automation for custom use-cases

> ## Repos

Author: Add a section for use-cases that will be supported

1. How DRA resources can be consumed by users of KubeVirt?
2. What changes are needed for existing device-plugin drivers to be DRA compatible?

This design focuses on part 1 of the problem. 

For allowing users to consume DRA devices, there are two main changes needed:
1. API changes
2. Driver Implementation

## API Changes

```
type VirtualMachineInstanceSpec struct {
    // ResourceClaims defines which ResourceClaims must be allocated
    // and reserved before the virt-launcher pod and hence the VMI is allowed to start. The resources
    // will be made available to the domain which consume them
    // by name.
    //
    // This is an alpha field and requires enabling the
    // DynamicResourceAllocation feature gate in kubernetes
    //  https://kubernetes.io/docs/concepts/scheduling-eviction/dynamic-resource-allocation/
    //
    // This field is immutable.
    //
    // +featureGate=DynamicResourceAllocation
    ResourceClaims []k8sv1.PodResourceClaim `json:"resourceClaims,omitempty"`
}

type GPU struct {
	// Name of the GPU device as exposed by a device plugin
	Name              string       `json:"name"`

> 	DeviceName        string       `json:"deviceName"`

Author: put deviceName and claim in an inline struct

	ClaimName         string       `json:"claimName"`
	VirtualGPUOptions *VGPUOptions `json:"virtualGPUOptions,omitempty"`
	// If specified, the virtual network interface address and its tag will be provided to the guest via config drive
	// +optional
	Tag   string `json:"tag,omitempty"`
	// Claim is the name of resource, defined in spec.resourceClaims,
	// that is used to provision this GPU device.
	//
	// This is an alpha field and requires enabling the
	// DynamicResourceAllocation feature gate.
	//
	// This field is immutable. It can only be set for containers.
	//
	// +featureGate=DynamicResourceAllocation
	// +optional
	Claim string `json:"claim,omitempty"`
}

type HostDevice struct {
	Name string `json:"name"`
	// DeviceName is the resource name of the host device exposed by a device plugin

> 	DeviceName string `json:"deviceName"`

Author: see the inline comment above

	// If specified, the virtual network interface address and its tag will be provided to the guest via config drive
	// +optional
	Tag string `json:"tag,omitempty"`
	// Claim is the name of resource, defined in spec.resourceClaims,
	// that is used to provision this host device.
	//
	// This is an alpha field and requires enabling the
	// DynamicResourceAllocation feature gate.
	//
	// This field is immutable. It can only be set for containers.
	//
	// +featureGate=DynamicResourceAllocation
	// +optional
	Claim string `json:"claim,omitempty"`

> }

Author: syncing with Varun offline, we decided that a vmi.Status section for mapping the resource claim name to the name used in vmi spec will be needed. This mapping will be used by virt-launcher to generate the env variable

      - name: a100-40C
        source:
          resourceClaimTemplateName: a100-40C-claim-template
      domain:
        devices:
          gpu:

>             name: vgpu

Author: ```suggestion
         -  name: vgpu
```
